lib/logstash/pluginmanager/util.rb

```diff
@@ -1,64 +1,101 @@
-    re = name ? /#{name}/i : //
-    specs = Gem::Specification.find_all{|spec| spec.name =~ re && req =~ spec.version}
-    specs.inject({}){|result, spec| result[spec.name_tuple] = spec; result}.values
+  # retrieve gem specs for all or specified name valid logstash plugins locally installed
+  # @param name [String] specific plugin name to find or nil for all plungins
+  # @return [Array<Gem::Specification>] all local logstash plugin gem specs
+  def self.find_plugins_gem_specs(name = nil)
+    specs = name ? Gem::Specification.find_all_by_name(name) : Gem::Specification.find_all
+    specs.select{|spec| is_logstash_plugin_gem_spec?(spec)}
   end
 
-  def self.ask_yesno(prompt)
-    while true
-      $stderr.puts ("#{prompt} [y/n]: ")
-      case $stdin.getc.downcase
-        when 'Y', 'y', 'j', 'J', 'yes' #j for Germans (Ja)
-          return true
-        when /\A[nN]o?\Z/ #n or no
-          break
-      end
-    end
+  # list of all locally installed plugins specs specified in the Gemfile.
+  # note that an installed plugin dependecies like codecs will not be listed, only those
+  # specifically listed in the Gemfile.
+  # @param gemfile [LogStash::Gemfile] the gemfile to validate against
+  # @return [Array<Gem::Specification>] list of plugin names
+  def self.all_installed_plugins_gem_specs(gemfile)
+    # we start form the installed gemspecs so we can verify the metadata for valid logstash plugin
+    # then filter out those not included in the Gemfile
+    find_plugins_gem_specs.select{|spec| !!gemfile.find(spec.name)}
   end
 
-end
+  # @param plugin [String] plugin name
+  # @param gemfile [LogStash::Gemfile] the gemfile to validate against
+  # @return [Boolean] true if the plugin is an installed logstash plugin and spefificed in the Gemfile
+  def self.is_installed_plugin?(plugin, gemfile)
+    !!gemfile.find(plugin) && !find_plugins_gem_specs(plugin).empty?
+  end
+
+  # @param plugin_list [Array] array of [plugin name, version] tuples
+  # @return [Array] array of [plugin name, version, ...] tuples when duplciate names have been merged and non duplicate version requirements added
+  def self.merge_duplicates(plugin_list)
+
+    # quick & dirty naive dedup for now
+    # TODO: properly merge versions requirements
+    plugin_list.uniq{|plugin| plugin.first}
+  end
 
-# This adds the "repo" element to the jar-dependencies DSL
-# allowing a gemspec to require a jar that exists in a custom
-# maven repository
-# Example:
-#   gemspec.requirements << "repo http://localhosty/repo"
-require 'maven/tools/dsl/project_gemspec'
-class Maven::Tools::DSL::ProjectGemspec
-  def repo(url)
-    @parent.repository(:id => url, :url => url)
+  # This adds the "repo" element to the jar-dependencies DSL
+  # allowing a gemspec to require a jar that exists in a custom
+  # maven repository
+  # Example:
+  #   gemspec.requirements << "repo http://localhosty/repo"
+  require 'maven/tools/dsl/project_gemspec'
+  class Maven::Tools::DSL::ProjectGemspec
```

Comment:
yeah I noticed that, but when reviewing this it was like 11:30pm and I was not about to embark on that journey :P